pkg/registry/apis/provisioning/read.go

```diff
@@ -60,38 +65,60 @@ func (s *readConnector) Connect(ctx context.Context, name string, opts runtime.O
 		return nil, fmt.Errorf("expected repository, but got %t", obj)
 	}
 
+	tokenSrc := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: repo.Spec.GitHub.Token},
+	)
+	tokenClient := oauth2.NewClient(ctx, tokenSrc)
+	githubClient := github.NewClient(tokenClient)
+
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		idx := strings.Index(r.URL.Path, "/"+name+"/read")
-		path := strings.TrimLeft(r.URL.Path[idx+len(name+"/read")+2:], "/")
-		if path == "" {
+		filePath := strings.TrimLeft(r.URL.Path[idx+len(name+"/read")+2:], "/")
+		if filePath == "" {
 			responder.Error(fmt.Errorf("missing path")) // TODO bad request
 			return
 		}
 		commit := r.URL.Query().Get("commit")
 
-		// TODO, actually read, and then validate
-		obj := &unstructured.Unstructured{
-			Object: map[string]interface{}{
-				"apiVersion": "dashboards.grafana.app/v????",
-				"metadata": metav1.ObjectMeta{
-					Name: "xxx",
-				},
-				"spec": map[string]interface{}{
-					"path":   path,
-					"commit": commit,
-					"repo":   repo.Name,
-					"type":   repo.Spec.Type,
-				},
-			},
+		owner, repoName, err := extractOwnerAndRepo(repo.Spec.GitHub.Repository)
+		if err != nil {
+			// TODO: handle
+			responder.Error(err)
+			return
+		}
+
+		content, _, _, err := githubClient.Repositories.GetContents(ctx, owner, repoName, filePath, &github.RepositoryContentGetOptions{
+			Ref: commit,
+		})
+		if err != nil {
+			// TODO: handle 404
+			// QUESTION: how to implement this?
```

Comment:
implement what? 404 handling?

Reply:
returning that status code. Let me change the TODO comments.